File: ts/src/values/stringValue.ts
import { CompareResult, IObjectValue, IValue } from "./ivalues";
import { NumberValue } from "./numberValue";

const keys: ReadonlyArray<string> = [ 'length' ];
export class StringValue implements IObjectValue
{
    public readonly value: string;

    constructor (value: string)
    {
        this.value = value;
    }

    public typename() { return 'string'; }
    public toString() { return this.value; }

    public compareTo(other: IValue): CompareResult
    {
        if (!(other instanceof StringValue))
        {
            return 1;
        }

        const diff = this.value.localeCompare(other.value);
        if (diff == 0) { return 0; }
        if (diff < 0) { return -1; }
        return 1;
    }

    public getIndex(index: number): number
    {
        if (index < 0)
        {
            return this.value.length + index;
        }

        return index;
    }

    public tryGetKey(key: string)
    {
        if (key === 'length')
        {
            return new NumberValue(this.value.length);
        }

        return undefined;
    }

    public objectKeys()
    {
        return keys;
    }
}

export function isStringValue(input: IValue | undefined): input is StringValue
{
    return input instanceof StringValue;
}

File: unity/lysithea-vm-unity/Assets/Scripts/LysitheaVM/Values/StringValue.cs
#nullable enable

using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Runtime.CompilerServices;

namespace LysitheaVM
{
    public struct StringValue : IObjectValue
    {
        #region Fields
        private static readonly IReadOnlyList<string> Keys = new [] { "length" };
        public IReadOnlyList<string> ObjectKeys => Keys;
        public string TypeName => "string";

        public readonly string Value;
        #endregion

        #region Constructor
        public StringValue(string value)
        {
            this.Value = value;
        }
        #endregion

        #region Methods
        public int CompareTo(IValue? other)
        {
            if (other == null) return 1;
            if (other is StringValue otherString)
            {
                return this.Value.CompareTo(otherString.Value);
            }

            return 1;
        }

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public int GetIndex(int index)
        {
            if (index < 0)
            {
                return this.Value.Length + index;
            }

            return index;
        }

        public override string ToString() => this.Value;
        public string ToStringSerialise() => '"' + StandardStringLibrary.EscapedString(this.Value) + '"';

        public bool TryGetKey(string key, [NotNullWhen(true)] out IValue? value)
        {
            if (key == "length")
            {
                value = new NumberValue(this.Value.Length);
                return true;
            }

            value = null;
            return false;
        }
        #endregion
    }
}